plugins/scale/wayfire/plugins/scale-transform.hpp

```diff
@@ -90,75 +100,103 @@ class scale_transformer_t : public wf::view_2D
-                [z_order] (const auto& other)
+        bool view_pad_shrunk  = false;
+        bool view_pad_grown   = false;
+        bool scale_pad_shrunk = false;
+        for (auto& p : overlays)
         {
-            return other->z_order >= z_order;
-        });
+            auto& ol = *(p.second);
+            if (ol.pre_hook)
+            {
+                padding old_view_pad  = ol.view_padding;
+                padding old_scale_pad = ol.scale_padding;
+                (*ol.pre_hook)();
+
+                if (ol.view_padding.any_smaller_than(old_view_pad))
+                {
+                    view_pad_shrunk = true;
+                } else if (old_view_pad.any_smaller_than(ol.view_padding))
+                {
+                    view_padding.extend(ol.view_padding);
+                    view_pad_grown = true;
+                }
+
+                if (ol.scale_padding.any_smaller_than(old_scale_pad))
+                {
+                    scale_pad_shrunk = true;
+                } else if (old_scale_pad.any_smaller_than(ol.scale_padding))
+                {
+                    scale_padding.extend(ol.scale_padding);
+                }
+            }
+        }
 
-        it = overlays.insert(it, std::make_unique<overlay>(ol, pad, z_order));
-        auto& new_ol = **it;
-        if (new_ol.pad_cb)
+        if (view_pad_shrunk || view_pad_grown)
         {
-            (*new_ol.pad_cb)(new_ol.pad, view->get_bounding_box(this));
-            this->pad.extend(new_ol.pad);
+            view->damage();
         }
 
-        view->damage();
-    }
+        if (view_pad_shrunk || scale_pad_shrunk)
+        {
+            recalculate_padding();
+        }
 
-    void add_overlay(std::shared_ptr<render_callback_t>&& ol, int z_order,
-        std::shared_ptr<padding_callback_t>& pad)
-    {
-        add_overlay(ol, z_order, pad);
+        if (view_pad_shrunk)
+        {
+            view->damage();
+        }
     }
 
-    void add_overlay(std::shared_ptr<render_callback_t>& ol, int z_order,
-        std::shared_ptr<padding_callback_t>&& pad = std::shared_ptr<padding_callback_t>())
+    /**
+     * Add a new overlay that is rendered after this transform.
+     *
+     * @param ol      The overlay object to be added.
+     * @param z_order Relative order; overlays are called in order.
+     */
+    void add_overlay(std::shared_ptr<overlay> ol, int z_order)
     {
-        add_overlay(ol, z_order, pad);
-    }
+        if (!ol)
+        {
+            return;
+        }
 
-    void add_overlay(std::shared_ptr<render_callback_t>&& ol, int z_order,
-        std::shared_ptr<padding_callback_t>&& pad = std::shared_ptr<padding_callback_t>())
-    {
-        add_overlay(ol, z_order, pad);
+        auto it = std::find_if(overlays.begin(), overlays.end(),
+            [z_order] (const auto& other)
+        {
+            return other.first >= z_order;
+        });
+
+        view_padding.extend(ol->view_padding);
```

Comment:
won't padding be updated by the prehook anyway?

Reply:
It would, but this way, there is no need to call `damage()` twice; also, this will work if this call happens between our pre-hooks and rendering (I'm not sure why that would happen, but it would be completely OK to do that, e.g. from another plugin's pre-hook)

Reply:
ok, sounds sensible.